flask/_compat.py

```diff
@@ -97,3 +97,29 @@ def __exit__(self, *args):
         BROKEN_PYPY_CTXMGR_EXIT = True
     except AssertionError:
         pass
+
+
+try:
+    from os import fspath
+except ImportError:
+    # Adapted from https://github.com/python/cpython/blob/3.7/Lib/os.py#L1029-L1058
+    def fspath(path):
+        if isinstance(path, (text_type, bytes)):
+            return path
+
+        path_type = type(path)
+        try:
+            path_repr = path_type.__fspath__(path)
+        except AttributeError:
+            if hasattr(path_type, '__fspath__'):
+                raise
+            else:
+                raise TypeError("expected str, bytes or os.PathLike object, "
+                                "not " + path_type.__name__)
+
+        if isinstance(path_repr, (text_type, bytes)):
+            return path_repr
+        else:
+            raise TypeError("expected {}.__fspath__() to return str or bytes, "
+                            "not {}".format(path_type.__name__,
+                                            type(path_repr).__name__))
```

Comment:
Under what circumstances would you have an object that supports the `__fspath__` protocol but not have `os.fspath` available? I'm hesitant to maintain backports unless they're essential. A simpler compat would be `def fspath(s): return str(path)`, which is what was done pre-3.6 anyway.

Reply:
Hmm, I guess it's unlikely? We could also just do this simpler version in the PEP: https://www.python.org/dev/peps/pep-0519/#backwards-compatibility

It is conceivable that a non python 3.6 library may implement a `__fspath__` as a forwards compatible thing.

Reply:
So the common library before `pathlib` was `path.py` and they seem to implement `__fspath__`: https://pypi.org/project/path.py/ https://github.com/jaraco/path.py/blob/master/path.py#L309

Reply:
Right, but nothing on a version of Python that doesn't support `os.fspath` would be taking advantage of that, they'd have to be passing `str(path)` around to all other built-in / library functions anyway.

Reply:
It's your call. I'm happy to do whatever here. I don't honestly have enough context to know what all the situations might be. I'd rather yank directly from the PEP for backwards compat if we don't want to port the version from python 3.6+ though since that's what's recommended. Coercing to `str()` always will change the type vs what is proposed would preserve the type.

Reply:
So I guess to be explicit:

```python
def fspath(path):
    return path.__fspath__() if hasattr(path, "__fspath__") else path
```